src/crystal/system/print_error.cr

```diff
@@ -1,15 +1,150 @@
+      {% end %}
+    end
+  end
+
+  # Minimal drop-in replacement for a C `printf` function. Yields successive
+  # `Bytes` to the block, which should do the actual printing.
+  #
+  # *format* only supports the `%(l|ll)?[dpsux]` format specifiers; more should
+  # be added only when we need them or an external library passes its own format
+  # string to here. *format* must also be null-terminated.
+  #
+  # Since this method may be called under low memory conditions or even with a
+  # corrupted heap, its implementation should be as low-level as possible,
+  # avoiding memory allocations.
+  #
+  # NOTE: C's `printf` is incompatible with Crystal's `sprintf`, because the
+  # latter does not support argument width specifiers nor `%p`.
+  def self.print_error(format, *args, &)
+    ptr, format_len = to_unsafe_string(format)
+    finish = ptr + format_len
+    arg_index = 0
+
+    while ptr < finish
+      next_percent = ptr
+      while next_percent < finish && !(next_percent.value === '%')
+        next_percent += 1
+      end
+      yield Slice.new(ptr, next_percent - ptr)
+
+      fmt_ptr = next_percent + 1
+      width = 0
+      if fmt_ptr.value === 'l'
+        width = 1
+        fmt_ptr += 1
+        if fmt_ptr.value === 'l'
+          width = 2
+          fmt_ptr += 1
+        end
+      end
+
+      break unless fmt_ptr < finish
+
+      case fmt_ptr.value
+      when 's'
+        if (arg = args[arg_index].as?(String | UInt8*)).nil?
+          yield "(???)".to_slice
+        else
+          yield Slice.new(*to_unsafe_string(arg))
+        end
+        arg_index += 1
+      when 'd'
+        if (arg = args[arg_index].as?(Int::Primitive)).nil?
+          yield "(???)".to_slice
+        else
+          to_unsafe_int(arg, 10, true, width) { |bytes| yield bytes }
+        end
+        arg_index += 1
+      when 'u'
+        if (arg = args[arg_index].as?(Int::Primitive)).nil?
+          yield "(???)".to_slice
+        else
+          to_unsafe_int(arg, 10, false, width) { |bytes| yield bytes }
+        end
+        arg_index += 1
+      when 'x'
+        if (arg = args[arg_index].as?(Int::Primitive)).nil?
+          yield "(???)".to_slice
+        else
+          to_unsafe_int(arg, 16, false, width) { |bytes| yield bytes }
+        end
+        arg_index += 1
+      when 'p'
+        if (arg = args[arg_index].as?(Void*)).nil?
+          yield "(???)".to_slice
+        else
+          # NOTE: MSVC uses `%X` rather than `0x%x`, we follow the latter on all platforms
+          yield "0x".to_slice
+          to_unsafe_int(arg.address, 16, false, 2) { |bytes| yield bytes }
+        end
+        arg_index += 1
+      else
+        yield Slice.new(next_percent, fmt_ptr + 1 - next_percent)
+      end
+
+      ptr = fmt_ptr + 1
+    end
+  end
+
+  private def self.to_unsafe_string(str)
+    if str.is_a?(UInt8*)
+      {str, LibC.strlen(str)}
+    else
+      str = str.to_s
+      {str.to_unsafe, str.bytesize}
+    end
+  end
```

Comment:
**nitpick:** I'm wondering if it wouldn't be better to reuse `Bytes` as a return type instead of a tuple which essentially reinvents the same type again. You don't get tuple deconstruction in the first line of `.print_error`, but that's not a big loss.